Request with encoded spaces are not handled correctly

The listener fails to handle URLs with spaces. The client sends the request properly encoded but the hybrid HTTP connection fails to parse the URL and throws an exception.  

```java

java.util.concurrent.CompletionException: java.lang.RuntimeException: java.net.URISyntaxException: Illegal character in path at index 68: sb://test.servicebus.windows.net/echo/api/contents/Untitled Folder?type=directory&_=1644956591273
	at java.base/java.util.concurrent.CompletableFuture.encodeThrowable(CompletableFuture.java:314)
	at java.base/java.util.concurrent.CompletableFuture.completeThrowable(CompletableFuture.java:319)
	at java.base/java.util.concurrent.CompletableFuture$AsyncRun.run(CompletableFuture.java:1739)
	at java.base/java.util.concurrent.CompletableFuture$AsyncRun.exec(CompletableFuture.java:1728)
	at java.base/java.util.concurrent.ForkJoinTask.doExec(ForkJoinTask.java:290)
	at java.base/java.util.concurrent.ForkJoinPool$WorkQueue.topLevelExec(ForkJoinPool.java:1020)
	at java.base/java.util.concurrent.ForkJoinPool.scan(ForkJoinPool.java:1656)
	at java.base/java.util.concurrent.ForkJoinPool.runWorker(ForkJoinPool.java:1594)
	at java.base/java.util.concurrent.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:183)
Caused by: java.lang.RuntimeException: java.net.URISyntaxException: Illegal character in path at index 68: sb://test.servicebus.windows.net/echo/api/contents/Untitled Folder?type=directory&_=1644956591273
	at com.microsoft.azure.relay.HybridHttpConnection.invokeRequestHandler(HybridHttpConnection.java:173)
	at com.microsoft.azure.relay.HybridHttpConnection.lambda$processFirstRequestAsync$2(HybridHttpConnection.java:115)
	at java.base/java.util.concurrent.CompletableFuture$AsyncRun.run(CompletableFuture.java:1736)
	... 6 common frames omitted
Caused by: java.net.URISyntaxException: Illegal character in path at index 68: sb://test.servicebus.windows.net/echo/api/contents/Untitled Folder?type=directory&_=1644956591273
	at java.base/java.net.URI$Parser.fail(URI.java:2913)
	at java.base/java.net.URI$Parser.checkChars(URI.java:3084)
	at java.base/java.net.URI$Parser.parseHierarchical(URI.java:3166)
	at java.base/java.net.URI$Parser.parse(URI.java:3114)
	at java.base/java.net.URI.<init>(URI.java:600)
	at com.microsoft.azure.relay.HybridHttpConnection.invokeRequestHandler(HybridHttpConnection.java:170)
	... 8 common frames omitted
```

--- a/src/main/java/com/microsoft/azure/relay/HybridHttpConnection.java
+++ b/src/main/java/com/microsoft/azure/relay/HybridHttpConnection.java
@@ -20,6 +20,7 @@ import javax.websocket.CloseReason;
 import javax.websocket.CloseReason.CloseCodes;
 
 import org.eclipse.jetty.http.HttpStatus;
+import org.eclipse.jetty.util.URIUtil;
 import org.json.JSONObject;
 
 class HybridHttpConnection implements RelayTraceSource {
@@ -164,11 +165,7 @@ class HybridHttpConnection implements RelayTraceSource {
 		URI requestUri = null;
 
 		try {
-			String listenerAddressStr = listenerAddress.toString();
-			requestTarget = requestTarget.replaceFirst(listenerAddress.getPath(), "");
-			requestUri = (listenerAddressStr.endsWith("/") || requestTarget.startsWith("/"))
-					? new URI(listenerAddressStr + requestTarget)
-					: new URI(listenerAddressStr + "/" + requestTarget);
+			requestUri = parseRequestUri(listenerAddress, requestTarget);
 		} catch (URISyntaxException e) {
 			throw new RuntimeException(e);
 		}
@@ -206,6 +203,23 @@ class HybridHttpConnection implements RelayTraceSource {
             listenerContext.getResponse().setStatusDescription("The listener RequestHandler has not been configured.");
 			listenerContext.getResponse().close();
 		}
+	}
+
+	private URI parseRequestUri(URI listenerAddress, String requestTarget) throws URISyntaxException {
+		URI requestUri;
+		String listenerAddressStr = listenerAddress.toString();
+		String requestTargetWithoutConnectionName = requestTarget.replaceFirst(listenerAddress.getPath(), "");
+
+		if (requestTargetWithoutConnectionName.startsWith("/")){
+			requestTargetWithoutConnectionName = requestTargetWithoutConnectionName.replaceFirst("/","");
+		}
+
+		String[] pathAndQuery = requestTargetWithoutConnectionName.split("\\?",2);
+		pathAndQuery[0] = URIUtil.encodePath(pathAndQuery[0]);
+
+		requestUri = new URI(listenerAddressStr + "/" +
+				String.join("?",pathAndQuery));
+		return requestUri;
 	}
 
 	private CompletableFuture<Void> sendResponseAsync(ListenerCommand.ResponseCommand responseCommand,

--- a/src/main/java/com/microsoft/azure/relay/HybridHttpConnection.java
+++ b/src/main/java/com/microsoft/azure/relay/HybridHttpConnection.java
@@ -165,7 +165,7 @@ class HybridHttpConnection implements RelayTraceSource {
 		URI requestUri = null;
 
 		try {
-			requestUri = parseRequestUri(listenerAddress, requestTarget);
+			requestUri = parseAndEncodeRequestUri(listenerAddress, requestTarget);
 		} catch (URISyntaxException e) {
 			throw new RuntimeException(e);
 		}
@@ -205,7 +205,7 @@ class HybridHttpConnection implements RelayTraceSource {
 		}
 	}
 
-	private URI parseRequestUri(URI listenerAddress, String requestTarget) throws URISyntaxException {
+	static URI parseAndEncodeRequestUri(URI listenerAddress, String requestTarget) throws URISyntaxException {
 		URI requestUri;
 		String listenerAddressStr = listenerAddress.toString();
 		String requestTargetWithoutConnectionName = requestTarget.replaceFirst(listenerAddress.getPath(), "");

--- a/src/test/java/com/microsoft/azure/relay/HybridConnectionListenerTest.java
+++ b/src/test/java/com/microsoft/azure/relay/HybridConnectionListenerTest.java
@@ -3,12 +3,14 @@ package com.microsoft.azure.relay;
 import static org.junit.Assert.*;
 
 import java.io.IOException;
+import java.io.InputStream;
 import java.io.OutputStream;
 import java.net.HttpURLConnection;
 import java.net.URI;
 import java.net.URISyntaxException;
 import java.net.URL;
 import java.nio.ByteBuffer;
+import java.nio.charset.StandardCharsets;
 import java.time.Duration;
 import java.util.Arrays;
 import java.util.HashMap;
@@ -322,6 +324,8 @@ public class HybridConnectionListenerTest {
 		CompletableFuture<Void> requestReceived = new CompletableFuture<Void>();
 		String extraPath = "/extra Path/extra  Path2";
 		String extraQuery = "queryKey=queryValue&queryKey2=queryValue;"; // mixing in a special char just for fun
+		String responseText = "{\"foo\":\"bar\"}";
+		int responseStatusCode = 200;
 
 		listener.setRequestHandler((context) -> {
 			try {
@@ -336,6 +340,11 @@ public class HybridConnectionListenerTest {
 						requestUri.getPath()
 				);
 				assertEquals("The query wasn't the expected value", extraQuery, requestUri.getQuery());
+
+				context.getResponse().setStatusCode(responseStatusCode);
+				OutputStream responseOutput = context.getResponse().getOutputStream();
+				responseOutput.write(responseText.getBytes(StandardCharsets.UTF_8));
+				responseOutput.close();
 
 				requestReceived.complete(null);
 			} catch (Throwable ex) {
@@ -357,10 +366,18 @@ public class HybridConnectionListenerTest {
 		HttpURLConnection conn = (HttpURLConnection)url.openConnection();
 		conn.setRequestMethod("GET");
 		conn.setRequestProperty("ServiceBusAuthorization", tokenString);
+		assertEquals("Response code is not the expected value",responseStatusCode, conn.getResponseCode());
+		assertEquals("Response message is not the expected value","OK", conn.getResponseMessage());
 
-		conn.getResponseCode();
+		byte[] body = new byte[responseText.length()];
+		((InputStream)conn.getContent()).read(body);
+		String receivedResponseText = new String(body, StandardCharsets.UTF_8);
+
+		assertEquals("Response body is not the expected value", responseText, receivedResponseText);
+
 		try {
 			assertNull(requestReceived.get(30, TimeUnit.SECONDS));
+
 		} catch (Exception e) {
 			fail(e.getMessage());
 		}

--- a/src/test/java/com/microsoft/azure/relay/HybridConnectionListenerTest.java
+++ b/src/test/java/com/microsoft/azure/relay/HybridConnectionListenerTest.java
@@ -22,6 +22,7 @@ import java.util.concurrent.atomic.AtomicInteger;
 import java.util.concurrent.atomic.AtomicReference;
 
 import org.eclipse.jetty.http.HttpStatus;
+import org.eclipse.jetty.util.URIUtil;
 import org.eclipse.jetty.websocket.api.StatusCode;
 import org.eclipse.jetty.websocket.api.UpgradeException;
 import org.junit.After;
@@ -315,7 +316,57 @@ public class HybridConnectionListenerTest {
 			fail(e.getMessage());
 		}
 	}
-	
+
+	@Test
+	public void parseRequestPathWithSpacesAndQueryTest() throws IOException {
+		CompletableFuture<Void> requestReceived = new CompletableFuture<Void>();
+		String extraPath = "/extra Path/extra  Path2";
+		String extraQuery = "queryKey=queryValue&queryKey2=queryValue;"; // mixing in a special char just for fun
+
+		listener.setRequestHandler((context) -> {
+			try {
+				assertNotNull("Listener should have received a valid http context.", context);
+				assertNotNull("Listener should have received a valid http request.", context.getRequest());
+
+				URI requestUri = context.getRequest().getUri();
+				assertNotNull("Listener should have a URI from request", requestUri);
+				assertEquals(
+						"The path wasn't the expected value",
+						"/" + TestUtil.ENTITY_PATH + extraPath,
+						requestUri.getPath()
+				);
+				assertEquals("The query wasn't the expected value", extraQuery, requestUri.getQuery());
+
+				requestReceived.complete(null);
+			} catch (Throwable ex) {
+				requestReceived.completeExceptionally(ex);
+			} finally {
+				context.getResponse().close();
+			}
+		});
+
+		listener.openAsync(Duration.ofSeconds(15)).join();
+		StringBuilder urlBuilder = new StringBuilder(TestUtil.RELAY_NAMESPACE_URI + TestUtil.ENTITY_PATH);
+		urlBuilder.replace(0, 5, "https://");
+		urlBuilder.append(URIUtil.encodePath(extraPath));
+		urlBuilder.append("?").append(extraQuery);
+
+		URL url = new URL(urlBuilder.toString());
+		String tokenString = tokenProvider.getTokenAsync(url.toString(), Duration.ofHours(1)).join().getToken();
+
+		HttpURLConnection conn = (HttpURLConnection)url.openConnection();
+		conn.setRequestMethod("GET");
+		conn.setRequestProperty("ServiceBusAuthorization", tokenString);
+
+		conn.getResponseCode();
+		try {
+			assertNull(requestReceived.get(30, TimeUnit.SECONDS));
+		} catch (Exception e) {
+			fail(e.getMessage());
+		}
+	}
+
+
 	@Test
 	public void listenerReconnectionTest() {
 		AtomicInteger handlerExecuted = new AtomicInteger(0);

